pageserver/src/tenant/timeline.rs

```diff
@@ -1749,37 +1527,33 @@ impl Timeline {
             // If so, rename_to_backup those files & replace their local layer with
             // a RemoteLayer in the layer map so that we re-download them on-demand.
             if let Some(local_layer) = local_layer {
-                let local_layer_path = local_layer
-                    .local_path()
-                    .expect("caller must ensure that local_layers only contains local layers");
-                ensure!(
-                    local_layer_path.exists(),
-                    "every layer from local_layers must exist on disk: {}",
-                    local_layer_path.display()
-                );
+                // FIXME: I removed an exists check which was not try_exists
 
-                let remote_size = remote_layer_metadata.file_size();
-                let metadata = local_layer_path.metadata().with_context(|| {
-                    format!(
-                        "get file size of local layer {}",
-                        local_layer_path.display()
-                    )
+                let path = local_layer.local_path();
+
+                // FIXME: this could be async, but that'd be much more horrible
+                // FIXME: should just load local layers as ResidentLayer but we need something
+                // other than what the {Delta,Image}LayerWriters use
+                let needs_download = local_layer.needs_download_blocking().with_context(|| {
+                    format!("check if local layer needs downloading {}", path.display())
                 })?;
-                let local_size = metadata.len();
-                if local_size != remote_size {
-                    warn!("removing local file {local_layer_path:?} because it has unexpected length {local_size}; length in remote index is {remote_size}");
-                    if let Err(err) = rename_to_backup(&local_layer_path) {
-                        assert!(local_layer_path.exists(), "we would leave the local_layer without a file if this does not hold: {}", local_layer_path.display());
-                        anyhow::bail!("could not rename file {local_layer_path:?}: {err:?}");
-                    } else {
-                        self.metrics.resident_physical_size_gauge.sub(local_size);
-                        corrupted_local_layers.push(local_layer);
-                        // fall-through to adding the remote layer
+
+                if let Some(reason) = needs_download {
```

Comment:
This is wrong, it will latch on to the initial local layer info, leading to test failures like https://neon-github-public-dev.s3.amazonaws.com/reports/pr-4938/5884521805/index.html#suites/f809933e0f6e3c9316f865c5a9d25004/f2d87cd7b8d25edd/

Christian also pointed this out to me in a call we had, or at least something in that direction.

Reply:
Scrolling a mile down and up, I see, I could just keep the current structure where proceed to delete the not really matching length files. Feels very hacky.

Thinking there should be one init, and nothing escapes before that.

Reply:
I've done the above hacky way.. Should probably clean up in a separate, unless done by Christian.

Reply:
Done #5074.